### PR TITLE
Add support for HTTP Range to `FileResponse`

### LINKED ISSUE #950: Support for HTTP range requests (for FileResponse)
I'm trying to embed an mp4 file on a page using the following HTML:

    <video controls width="600">
        <source src="/media/video.mp4" type="video/mp4">
    </video>

The video file is being served by a Starlette `FileResponse`.

I'm getting this error in Safari:

localhost_8001_live-photos_and_Real-time_HTML_Editor

It looks to me like Safari is trying to make an HTTP range request in order to stream the video - but Starlette doesn't support that option.

I tried adding `accept-ranges: none` as a response header but that didn't seem to fix the problem.

So... it would be great if Starlette could handle range requests so you could use it to serve video files to Safari!

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/950">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/950/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Kludex**
- Closes https://github.com/encode/starlette/issues/950

Inspired by Baize and copied from my own implementation: https://github.com/Kludex/file-response .

### CODE REVIEW

**Review by abersheeran — APPROVED** [review]
hahaha, I like this.

**Comment by trim21** [comment]
We should pre-compile regex pattern. Althrough `re` has a built-in cache, but it has size limit and for project using many regex it still may be invalidated and pattern get re-compiled.

**Comment by Kludex** [comment]
> We should pre-compile regex pattern. Althrough `re` has a built-in cache, but it has size limit and for project using many regex it still may be invalidated and pattern get re-compiled.

PR welcome. 👍

File: starlette/responses.py [hunk 1]
```diff
@@ -342,5 +379,154 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+                    await send({"type": "http.response.body", "body": chunk, "more_body": more_body})
+
+    async def _handle_multiple_ranges(
+        self,
+        send: Send,
+        ranges: list[tuple[int, int]],
+        file_size: int,
+        send_header_only: bool,
+    ) -> None:
+        boundary = "".join(random_choices("abcdefghijklmnopqrstuvwxyz0123456789", k=13))
```
Comment (graingert) [inline starlette/responses.py:408]: I'm a bit concerned that this uses predictable randomness

I'd be tempted to go for 19 characters using SystemRandom.choices to get 96 bits of entropy

Comment (trim21) [reply starlette/responses.py:408]: why `join(random_choices)` anyway? 

It's also slower than `secrets.token_hex(6/7/13)` unless 13 has some special meaning here.

Comment (Kludex) [reply starlette/responses.py:408]: Do other frameworks do something different from this?

Would you like to open a PR for it?

File: starlette/responses.py [hunk 2]
```diff
@@ -342,5 +379,154 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
...
+        self.headers["content-range"] = f"multipart/byteranges; boundary={boundary}"
+        self.headers["content-length"] = str(content_length)
+        await send({"type": "http.response.start", "status": 206, "headers": self.raw_headers})
+        if send_header_only:
+            await send({"type": "http.response.body", "body": b"", "more_body": False})
+        else:
+            async with await anyio.open_file(self.path, mode="rb") as file:
+                for start, end in ranges:
+                    await file.seek(start)
+                    chunk = await file.read(min(self.chunk_size, end - start))
```
Comment (frostming) [inline starlette/responses.py:421]: Does it send all content within the range?

Comment (Kludex) [reply starlette/responses.py:421]: I'm not sure I understand the question... Could you rephrase it?

File: starlette/responses.py (hunk 1 above)
Comment (abersheeran) [reply starlette/responses.py:408]: > 13 has some special meaning here

Nothing special. It's just a random number I picked, and the spec doesn't make any specific requirements on this length.

Comment (abersheeran) [reply starlette/responses.py:408]: > I'm a bit concerned that this uses predictable randomness
> 
> I'd be tempted to go for 19 characters using SystemRandom.choices to get 96 bits of entropy

This is predicted to be secure...it does not serve any cryptographic purpose.

Comment (graingert) [reply starlette/responses.py:408]: Urllib3 does this https://github.com/urllib3/urllib3/blob/2458bfcd3dacdf6c196e98d077fc6bb02a5fc1df/src/urllib3/filepost.py#L26 (it's for form data but same sort of concept)

nginx is just totally weird https://github.com/nginx/nginx/blob/b1e07409b1071564e8dd8c70436cba5da19a1575/src/core/ngx_file.c#L17 it uses a random number generator based on successive additions of  123456

Apache is bizarre also, it's using predictable randomness, but it's a constant defined each time config is loaded https://github.com/apache/httpd/blob/64deb516d6a3b8d872dab9d27a1f0cba17548190/modules/http/http_core.c#L350

File: starlette/responses.py (hunk 2 above)
Comment (frostming) [reply starlette/responses.py:421]: IIUC, this statement may send chunk with self.chunk_size, not the full content between `end - start`, and I didn't find similar logic to that in `handle_single_range`(pasted below) to loop until no more body to send：

```python
while more_body:
    chunk = await file.read(min(self.chunk_size, end - start))
    start += len(chunk)
    more_body = len(chunk) == self.chunk_size and start < end
    await send({"type": "http.response.body", "body": chunk, "more_body": more_body})
```

File: starlette/responses.py (hunk 1 above)
Comment (frostming) [reply starlette/responses.py:408]: Yes, urllib3 implementation is just equivalent to `secrets.token_hex()`

File: starlette/responses.py (hunk 2 above)
Comment (Kludex) [reply starlette/responses.py:421]: I think we may benefit from some comments/docstrings around here. But I don't see anything wrong?

Could you try to come up with a test scenario that you think it would be problematic here?

This is a basic test for multiple ranges (to help):

```py
def test_file_response_merge_ranges(file_response_client: TestClient) -> None:
    response = file_response_client.get("/", headers={"Range": "bytes=0-100, 50-200"})
    assert response.status_code == 206
    assert response.headers["content-length"] == "201"
    assert response.headers["content-range"] == f"bytes 0-200/{len(README.encode('utf8'))}"
```

File: starlette/responses.py (hunk 1 above)
Comment (graingert) [reply starlette/responses.py:408]: Yeah I'd probably go for secrets.token_hex()

File: starlette/responses.py (hunk 2 above)
Comment (frostming) [reply starlette/responses.py:421]: Yes, the above test works but needs to adjust the `chunk_size` to less than 100. 

I thought you would immediately see my points. This statement

```python
chunk = await file.read(min(self.chunk_size, end - start))
```
may not send full content if the chunk_size is smaller, and you didn't send the rest anywhere in this branch.

Comment (Kludex) [reply starlette/responses.py:421]: Oh... I see your point.

Comment (Kludex) [reply starlette/responses.py:421]: PR is welcome here, otherwise I'll work on it later.

Comment (Kludex) [reply starlette/responses.py:421]: Thanks @frostming 🙏
